### ISSUE TITLE
database/sql: provide a way for a driver to get its own driver.Rows struct from sql.Row and sql.Rows

### ISSUE THREAD

**Issue by gopherbot** [golang/go#5606]
by **arnehormann**:

<pre>What is the expected output?

There should be a way to access the full information a driver has for result columns.
Workarounds are brittle and slow. They depend on the struct layout, require reflect or
unsafe and involve some duplication.

What do you see instead?

I can only get the names by calling Columns(). The rest of the result metadata can not
be retrieved.

Please provide any additional information below.

I posted a proposal on go-nuts:
<a href="https://groups.google.com/forum/#">https://groups.google.com/forum/#</a>!topic/golang-nuts/2aLctcVyp6Q

In short, I propose to add a new interface to database/sql/driver that passes a callback
function to the driver.
The callback is returned after a type assertion in database/sql.Register.
It can be used to get the driver.Rows implementation given database/sql/Row[s].
The native Rows implementation by the driver is returned as interface{}.

Given this function, a driver can decide to expose more detailed metadata for the result
columns by implementing the interface and providing an exported function that retrieves
the metadata from *Row or *Rows.</pre>

**Comment by matrixik** [comment]
Any updates?

**Comment by h4ck3rm1k3** [comment]
I am also interested in this. Just found this thread after trying to get access to the native row implementation. It seems to me that the hack to https://github.com/arnehormann/sqlinternals/blob/e3f9da33ba6e09e023c4b1d32ea73b39d3132d59/unsafe.go#L102 inspect the value is a bad solution. I would rather fork the database/sql and add in more interfaces.
Am going to take a look at this module if it can be used.
https://github.com/flynn/go-sql

**Comment by johto** [comment]
I find the interface a bit weird. My first instinct would have been for the drivers to implement a new method in the database/sql/driver.Rows object they're returning.  This method would return an interface{} which the user of database/sql would then assert to structs or interfaces defined by the driver.  This could be nicer for backwards compatibility as well, since you could check whether the returned object implements a certain interface added later in the driver's life cycle.

**Comment by h4ck3rm1k3** [comment]
Well for my purposes, I found github.com/kdar/dbtogo to generate the go schema for my tables. With that schema generated I know the types. Then I am using upperdb https://upper.io/db as a replacement for the standard go driver. It all works out OK.

**Comment by kardianos** [comment]
This looks duplicate of https://github.com/golang/go/issues/16652 . Recommend closing this issue.
